services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/ReceiveSqsMessageHelper.java

```diff
@@ -0,0 +1,163 @@
+    private volatile long visibilityDeadlineNano;
+
+    public ReceiveSqsMessageHelper(String queueUrl,
+                                   SqsAsyncClient asyncClient,
+                                   Duration visibilityTimeout,
+                                   ResponseBatchConfiguration config) {
+        this.queueUrl = queueUrl;
+        this.asyncClient = asyncClient;
+        this.visibilityTimeout = visibilityTimeout;
+        this.config = config;
+    }
+
+    public CompletableFuture<ReceiveSqsMessageHelper> asyncReceiveMessage() {
+        ReceiveMessageRequest.Builder request =
+            ReceiveMessageRequest.builder()
+                                 .queueUrl(queueUrl)
+                                 .maxNumberOfMessages(config.maxBatchItems())
+                                 .messageSystemAttributeNames(config.messageSystemAttributeNames())
+                                 .messageAttributeNames(config.receiveMessageAttributeNames());
+
+        request.visibilityTimeout(NumericUtils.saturatedCast(this.visibilityTimeout.getSeconds()));
+
+        if (config.longPollWaitTimeout() != null) {
+            request.waitTimeSeconds(NumericUtils.saturatedCast(config.longPollWaitTimeout().getSeconds()));
+        }
+        try {
+            return asyncClient.receiveMessage(request.build())
+                              .handle((response, throwable) -> {
+                                  if (throwable != null) {
+                                      setException(throwable);
+                                  } else {
+                                      messages.addAll(response.messages());
+                                  }
+                                  return this;
+                              });
+        } finally {
+            visibilityDeadlineNano = System.nanoTime() + visibilityTimeout.toNanos();
+        }
+
+    }
+
+    public boolean isEmpty() {
+        return messages == null || messages.isEmpty();
+    }
+
+    public Throwable getException() {
+        return exception;
+    }
+
+    public void setException(Throwable exception) {
+        this.exception = exception;
+    }
+
+    public Message removeMessage() {
+        if (isExpired()) {
+            clear();
+            return null;
+        }
+        return messages.poll();
+    }
+
+    private boolean isExpired() {
+        return System.nanoTime() > visibilityDeadlineNano;
+    }
+
+    public void clear() {
+        if (!isEmpty()) {
+            Optional.ofNullable(nackMessages())
+                    .ifPresent(future -> future.whenComplete((r, t) -> {
+                        // Logging an error is sufficient here as this is an asynchronous cleanup activity,
+                        // and there are no dependent tasks waiting for its completion.
+                        if (t != null) {
+                            log.error(() -> "Could not change visibility for queue " + queueUrl, t);
```

Comment:
Should this be warn instead? Do customers need to retry this by themselves? Can we add more instructions here? Example: https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/GenericMultipartHelper.java#L121

Reply:
Yeap, made it to warn.
No customer need not retry , the SQS will make those messages visible after visibilty timeout is over